# Long Read Sequencing - Assembler

This Snakemake workflow facilitates the hybrid assembly of long-read datasets generated by Oxford Nanopore Technologies (ONT) and Pacific Biosciences (PacBio) platforms. Designed for scalability and flexibility, it is an ideal solution to assemble near-complete genomes with minimal number of gaps.

In addition to genome assembly, the workflow supports the phasing and annotation of user-specified regions, such as highly heterozygous immune-related haplotypes. By utilizing the high contiguity of paternal and maternal contigs produced by hifiasm, and integrating a user-provided reference sequence database, it enables accurate annotation of these regions. 

## Installation Guide

To download the pipeline, the repository can be cloned using the following command:

    git clone git@github.com:BPRC-Bioinfo/LRS-Assembler.git
    cd ./LRS-Assembler

Then install the pipeline dependencies in a new enviroment:

    conda env create -f envs/LRS-assembler.yaml
    conda activate lrs_pipe

## Usage Guide

Before running LRS-Assembler, the user has to provide the location of the input data files and define the analysis variables in the configuration file. This file should specify the scientific name of the species studied and, optionally, include details on the flanking genes of the regions of interest and the location of the reference library. If necessary, the user can also adjust the settings for LiftOff, BUSCO and minimap2.

### Updating User-specific Variables

The configuration file, located at ```configs/run-config.yaml```, must be updated with user-specific variables before launching LRS-Assembler. Various settings can be modified, as outlined below.

A config file looks like:
```
species: "Macaca mulatta"
reference:
  GCF_003339765.1:
    accession_number: "GCF_003339765.1"
    genome: "/path/to/local/reference/genomic.fna"
    gff: "/path/to/local.gff"
    chr_info: "/path/to/local/chr_info_file"
region:
  KIR:
    left_flank: "FCAR"
    right_flank: "LILRA6"
    library: "references/mamu_kir_gen_2501.fasta" 
    minimap2: "-cx splice:hq -G16k"
    blast: "-word_size 7"
  MHC:
    left_flank: "KIFC1"
    right_flank: "GABBR1"
    library: "references/mamu_mhc_gen_2501.fasta" 
    minimap2: "-cx splice:hq -G16k "
    blast: "-word_size 7"
busco: "primates_odb10"
nanopore:
  Sample_1:
    - "/path/to/local/Sample_1/nanopore/fastq1"
  Sample_2:
    - "/path/to/local/Sample_2/nanopore/fastq1"
pacbio:
  Sample_1:
    - "/path/to/local/Sample_1/pacbio/bam1"
  Sample_2:
    - "/path/to/local/Sample_2/pacbio/bam1"
```
**Species**
Users should specify the scientific name of the species being studied. 

**Reference**
A reference sequence must be specified, either by providing an accession number (e.g., GCF_003339765.1) or by providing the location of a local genome assembly and its annotation file (gff). 

**Region (Optional)**
Users can define one or more regions of interest, which are identified by their flanking genes. Flanking genes are those located adjacent to the region of interest, and the provided gene names should match those listed in the NCBI database for the specified species. 

To annotate the specified regions, a reference database must be specified in the configuration file. The choice of the minimap2 command depends on whether the reference database contains genomic or transcriptomic sequences. For mapping transcriptomic data to genomic assemblies, the ```splice:hq``` option should be used. For genomic reference databases, options such as ```-ax asm5``` or ```-ax asm10``` are recommended.

**BUSCO**
To evaluate the completeness of the genome assembly, a BUSCO analysis is performed. This method relies on evolutionarily informed expectations of near-universal single-copy orthologs to assess genome completeness. Various lineage-specific datasets can be used for this assessment, which are available at: https://busco.ezlab.org/list_of_lineages.html.

**Nanopore** and **PacBio**
The directory path containing ONT or PacBio data should be specified for each sample individually. Multiple samples can be assembled in parallel. The ONT and PacBio reads do not need to be combined into a single file prior to starting the pipeline.

### Run The Pipeline

To execute the pipeline, ensure that the input data is placed in the specified directory and the variables in the configuration file are properly defined. Then, run the following command:

```
snakemake --cores <number_of_cores> -s scripts/Snakefile --use-conda
```

For a clean run, use:

```
snakemake --cores <number_of_cores> -s scripts/Snakefile --use-conda -q
```

## Output

For each sample, a set of output files is generated, organized as outlined below:

```
.
└── results/
    └── sample_name/
        ├── hifiasm/
        │   └── All data generated by hifiasm
        ├── info/
        │   └── Different files containing information on the assembly, including an assembly overview in a HTML report.
        ├── intermediates/
        │   └── Intermediate working files
        ├── raws/
        │   └── Size-filtered (5000 bp) raw HiFi and ONT reads
        └── scaffolds/
            └── All final scaffolds generated by LRS-Assembler. Also intermediate scaffolds generated by ntLink and Ragtag are saved.
```

## Sample Data

The pipeline has been developed and validated using HiFi and ONT long-read datasets derived from rhesus macaque of Indian origin and human samples HG01109, HG02080, and HG02723 from the Human Pangenome Reference Consortium (HPRC). The rhesus macaque data has been submitted to the European Nucleotide Archive (ENA, https://www.ebi.ac.uk/ena/browser/home) under project number PRJEB84466. 

## Troubleshooting

If you encounter any issues with genome assembly or RIO extraction and annotation, please feel free to contact us at le@bprc.nl or bruijnesteijn@bprc.nl, or report the issue directly in this repository. The pipeline is under continuous development to improve genome assemblies and enhance haplotype annotations.

## Cite

The manuscript detailing LRS-Assembler is currently in preparation.
